### PR TITLE
__await__ magic method and async/await

### PR DESCRIPTION
**Pull request by WarriorOfWire**
This change:

* adds `__await__()` support as the backing implementation hook for CircuitPython async/await noun/verb pair.
* enables async/await on lots of boards.
* **removes all special cases** of `if sys.implementation.name in ('micropython', 'circuitpython'):` from micropython async tests.  These tests now behave like CPython in CircuitPython.

By using `__await__()` rather than `__iter__()` (uPython uses iter), CircuitPython is more amenable to library interoperation with CPython.  Users (and crucially, library developers!) can implement non-blocking APIs for complex devices in pure Python without writing state machines and saddling their application code with polling machines.

This is syntax-only.  Without a task scheduler, one can use this language infrastructure via the `send()` method on the coroutine-generators though that is not the intended usage pattern.

Today there is an example task scheduler implementation based on a standard python `asyncio` usage pattern.  It is different from asyncio in material ways but it provides an easy way to schedule recurring tasks and asynchronously drive disparate tasks without a `loop()` method.  https://github.com/WarriorOfWire/tasko

This has used practically on my personal projects for months but has not achieved wide adoption through my personal repository.  The error text is typically close to what CPython outputs and most cases of invalid syntax are reported similarly.  It is expected that additional errata exist but the discovery and prioritization of resolution of the same should be shared going forward.

### CODE REVIEW

File: ports/atmel-samd/boards/arduino_mkrzero/mpconfigboard.mk [hunk 1]
```diff
@@ -11,3 +11,6 @@ LONGINT_IMPL = NONE
 CIRCUITPY_FULL_BUILD = 0
 
 SUPEROPT_GC = 0
+
+MICROPY_PY_ASYNC_AWAIT = 0
+
```
Comment (jepler) [inline ports/atmel-samd/boards/arduino_mkrzero/mpconfigboard.mk:16]: To placate the so-called "pre-commit hook", these added trailing newlines must be deleted.

Would the total number of changes be smaller if you made CIRCUITPY_FULL_BUILD = 0 imply MICROPY_PY_ASYNC_AWAIT = 0?

File: py/circuitpy_mpconfig.mk [hunk 2]
```diff
@@ -32,6 +32,8 @@
 CIRCUITPY_FULL_BUILD ?= 1
 CFLAGS += -DCIRCUITPY_FULL_BUILD=$(CIRCUITPY_FULL_BUILD)
 
+MICROPY_PY_ASYNC_AWAIT ?= $(CIRCUITPY_FULL_BUILD)
+CFLAGS += -DMICROPY_PY_ASYNC_AWAIT=$(MICROPY_PY_ASYNC_AWAIT)
```
Comment (jepler) [inline py/circuitpy_mpconfig.mk:36]: In fact isn't that what this line is doing?

File: py/circuitpy_mpconfig.mk [hunk 3]
```diff
@@ -290,3 +292,6 @@ CFLAGS += -DCIRCUITPY_WIFI=$(CIRCUITPY_WIFI)
 # Enabled micropython.native decorator (experimental)
 CIRCUITPY_ENABLE_MPY_NATIVE ?= 0
 CFLAGS += -DCIRCUITPY_ENABLE_MPY_NATIVE=$(CIRCUITPY_ENABLE_MPY_NATIVE)
+
+# async/await language keyword support
+MICROPY_PY_ASYNC_AWAIT ?= $(CIRCUITPY_FULL_BUILD)
```
Comment (jepler) [inline py/circuitpy_mpconfig.mk:297]: why is this repeated twice?  am I overlooking a difference?

File: py/compile.c [hunk 4]
```diff
@@ -40,6 +40,8 @@
 
 #if MICROPY_ENABLE_COMPILER
 
+#define DEBUG_PRINT(...) mp_printf(&mp_plat_print __VA_OPT__(,) __VA_ARGS__)
```
Comment (jepler) [inline py/compile.c:43]: Please remove

**Review by jepler — CHANGES_REQUESTED** [review]
Thanks for this, I'm very excited.  Particularly I'm happy to see the identical code that can be written in Python3 and in CircuitPython with these changes.

File: py/circuitpy_mpconfig.mk (hunk 3 above)
Comment (WarriorOfWire) [reply py/circuitpy_mpconfig.mk:297]: goofed that right up.

File: py/compile.c (hunk 4 above)
Comment (WarriorOfWire) [reply py/compile.c:43]: yep, thanks.

**Review by jepler — COMMENTED** [review]
Thank you, this looks much better.  Requesting Scott's review too.

**Review by jepler — APPROVED** [review]

**Review by tannewt — APPROVED** [review]
Very cool! Thank you for all of this work.

Have you thought about what hardware and library APIs could benefit from this?

**Comment by WarriorOfWire** [comment]
> 
> 
> Very cool! Thank you for all of this work.
> 
> Have you thought about what hardware and library APIs could benefit from this?

Yes, I can't wait wait to make a context manager for communication resources, e.g., `SPI`.
Devices like `BME680` take wall clock time to generate their results.  I want to see library developers produce APIs that accept a context manager that provides access to a bus object.  That context manager would bundle a chip select pin and the SPI bus together and provide the bus when nothing else happens to be using it at the particular moment.

```
class BME680:
  [...]
  async def read_sensor(self, async_bus_provider):
    # TODO: Add a context manager for `self` (a mutex) to make sure we're only doing 1 sensor thing at a time
    async with async_bus_provider as bus:
      self.start_measurement(bus)  # Completes quickly or you could make this async as well (e.g., for doing DMA before releasing the bus provider)
    # Release bus while the BME680 does its thing.  We'll check back later.
    # Other devices can use the SPI bus via the context manager.
    await loop.sleep(time_it_takes_for_bme680_to_sense)
    async with async_bus_provider as bus:
      sensor_state = self.read_measurement(bus)
    return sensor_state
```

Adding library support for this pattern allows synchronization to happen at the point of demand rather than at a top level workflow.  Users of this api can do like `sensor_data = await bme.read_sensor( bus_provider_for_cs14 )` and while the BME680 is doing that, their other stuff (like displayio or other I2C sensors or even other SPI devices) is all free to execute as soon as the processor has cycles.

In my own user interface application I see frequent sleeps for 5.8 milliseconds as I have a 100hz loop checking my rotary button for input.  In the future I would like to see interrupt-based sleep state in addition to this polling model - though the polling model is already a huge improvement in ergonomics for complex applications.
The interrupt usage would look like
```
interrupt_handle = Interrupt( some_pin )
task = a_coroutine_function
loop.register_interrupt( interrupt_handle, a_function )
```

The event loop would watch for the pin to be interrupted and add an instance of `a_coroutine_function` to the active coroutines to be run later.  The interrupt would be handled in C and would simply (I think) increment a counter (In no case would a python stack be introduced to an interrupt's execution context).  TBD how to efficiently indicate to a sleeping event loop that it should wake (perhaps inserting a WakeException that the loop could handle), but in the interim the event loop could poll its Interrupts on some schedule.  Without interrupts, polling is still pretty easy to do with `await event_loop.sleep(however_long_you_need_to)`

**Comment by siddacious** [comment]
@WarriorOfWire Thanks! This is great to see. I'm excited to start using it :)

That said, any pointers on reading material on how to use it would be appreciated as while I'm somewhat familiar with the concepts, I'm not quite sure where to start.

**Comment by WarriorOfWire** [comment]
@siddacious yeah, if you just want to write async/await code you should take a look at tasko https://github.com/WarriorOfWire/tasko

If you want to implement your own event loop from scratch then grab a coffee and follow along with David Beazley https://youtu.be/Y4Gt3Xjd7G8

In fact, watching that video might be the best recommendation on the topic I have in general.

**Comment by siddacious** [comment]
@WarriorOfWire perfect, thanks :)

**Comment by ErlendFj** [comment]
I am on a steep learning curve on CircuitPython and now I need to understand how to get async/await to work. First obstacle is I get: no module named 'asyncio'. 
I am on 6.3.0. Any hints to where I can read up on this?

**Comment by dhalbert** [comment]
@ErlendFj `asyncio` is not yet available, though we are looking at event and concurrency support in more depth for 8.0.0. However, `async/await` is turned on for most boards, and there are several libraries already available. See this forum post: https://forums.adafruit.com/viewtopic.php?f=60&t=183355

**Comment by ErlendFj** [comment]
Thanks for the swift answer. I'll try with yield or asynccp and see if I
can sort it out.

**Comment by ErlendFj** [comment]
sorry - the insert code did not work as wanted

**Comment by jepler** [comment]
I corrected the code embed.  However, a better place to seek help using CircuitPython is [our Discord](https://adafru.it/discord) in #help-with-circuitpython or the [Adafruit Forum](https://forums.adafruit.com/index.php). Since you're using a third party library (asynccp) you may also wish to inquire in whatever support forum they suggest.

**Comment by WarriorOfWire** [comment]
I'd suggest you should separate your app data from your app actions. Something like this:

```python
class AppState:
    def __init__(self):
        self.now_audio = 'Hei.wav'
        self.now_colour = BLACK

async def scan_NFC():
  # This is an unfortunate expression. Is there not a start_read_passive_target() and is_read_passive_target_done()? Because then you could actually read from this and not block for up to 100 milliseconds like you're doing here.
  uid = pn532.read_passive_target(timeout=0.1)
  print(".", end="")
  if uid is not None:
    text_ID = ''  # This wasn't storing to app_state so I trimmed it from app_state. Don't keep useless state!
    for e in uid:
      text_ID += '' + str(e)
    print(text_ID)

async def play_audio(app_state):
    print("s", end="")
    if self.now_audio != '':  
       audio.play(audiocore.WaveFile(open(app_state.now_audio, "rb")))

async def toggle_colour(app_state):
    print("t", end="")
    if app_state.now_colour is BLACK:
      pixels[0] = BLUE
      pixels.show()
      app_state.now_colour = BLUE
    elif app_state.now_colour is BLUE: 
      pixels[0] = BLACK
      pixels.show()
      app_state.now_colour = BLACK

def run():
    app_state = AppState()

    asynccp.schedule(frequency=1, coroutine_function=app.scan_NFC, app_state)
    asynccp.schedule(frequency=0.2, coroutine_function=app.play_audio, app_state)
    asynccp.schedule(frequency=0.1, coroutine_function=app.toggle_colour, app_state)
    asynccp.run()

if __name__ == '__main__':
    run()
```

**Comment by ErlendFj** [comment]
@jepler, @WarriorOfWire, thanks for your help. I shall go to discord, but since asynccp is not part of basic circuitpython I thought i'd risk a question here. I struggle to get my head around class/objects - in this instance I don't understand how the state variables work. I have got the code up and running through, and the different functions 'communicate' with each other by means og global variables. I'm sure that is the bad way to do it, but so far as I do not understand better, that is the way I have to do it. Agree about the pn532.read_passive_target(timeout=0.1) behaving as a blocker. I have squeezed the time down to 0.05s though. To get a start_read & is_read option I will have to figure out commands directly to the NFC chip. Apologize for misusing git for solving my problems.

**Pull request merged**